overlord/state/change.go

```diff
@@ -282,32 +291,119 @@ func init() {
 	}
 }
 
+func (c *Change) isTaskWaiting(visited map[string]taskWaitComputeStatus, t *Task, deps []*Task) bool {
+	taskID := t.ID()
+	computeStatus := visited[taskID]
+	switch computeStatus {
+	case taskWaitStatusComputing:
+		// Cyclic dependency detected, return false to short-circuit.
+		return false
+	case taskWaitStatusWaiting, taskWaitStatusNotWaiting:
+		return computeStatus == taskWaitStatusWaiting
+	}
+	visited[taskID] = taskWaitStatusComputing
+
+	var isWaiting bool
+depscheck:
+	for _, wt := range deps {
+		switch wt.Status() {
+		case WaitStatus:
+			isWaiting = true
+		// For 'Do' and 'Undo' we have to check whether the task is waiting
+		// for any dependencies. The logic is the same, but the set of tasks
+		// varies.
+		case DoStatus:
+			isWaiting = c.isTaskWaiting(visited, wt, wt.WaitTasks())
+			if !isWaiting {
+				// Cancel early if we detect something is runnable.
+				break depscheck
+			}
+		case UndoStatus:
+			isWaiting = c.isTaskWaiting(visited, wt, wt.HaltTasks())
+			if !isWaiting {
+				// Cancel early if we detect something is runnable.
+				break depscheck
+			}
+		}
```

Comment:
Shouldn't this be resetting `isWaiting` to `false` if a task in another status is observed?

```suggestion
        default:
            isWaiting = false
		}
```

If this processes a task in `WaitStatus` and then another in `DoingStatus`, it will return `true`. Or am I misunderstanding what this is supposed to do? I wrote a simple test to trigger this:
```
func (cs *changeSuite) TestExample(c *C) {
	st := state.New(nil)
	st.Lock()
	defer st.Unlock()

	chg := st.NewChange("change", "...")

	t1 := st.NewTask("task1", "...")
	t2 := st.NewTask("task2", "...")
	t3 := st.NewTask("task3", "...")
	t1.WaitFor(t2)
	t1.WaitFor(t3)

	chg.AddTask(t1)
	chg.AddTask(t2)
	chg.AddTask(t3)

	t2.SetToWait(state.DoneStatus)
	t3.SetStatus(state.DoingStatus)
	c.Check(chg.Status(), Equals, state.DoingStatus)
}
```

This fails with:
```
    c.Check(chg.Status(), Equals, state.DoingStatus)
... obtained state.Status = 10 ("Wait")
... expected state.Status = 3 ("Doing")
```

Reply:
The test example was really great, and exposed some incorrect logic I had leftover from my original implementation. Thanks for catching this! I added a default case also